tsdb/index/tsi1/partition.go

```diff
@@ -1119,20 +1151,29 @@ func (p *Partition) compactToLevel(files []*IndexFile, level int, interrupt <-ch
 	}
 
 	// Obtain lock to swap in index file and write manifest.
-	if err := func() error {
+	if err := func() (rErr error) {
 		p.mu.Lock()
 		defer p.mu.Unlock()
 
 		// Replace previous files with new index file.
-		p.fileSet = p.fileSet.MustReplace(IndexFiles(files).Files(), file)
+		newFileSet := p.fileSet.MustReplace(IndexFiles(files).Files(), file)
 
 		// Write new manifest.
-		manifestSize, err := p.Manifest().Write()
+		manifestSize, err := p.manifest(newFileSet).Write()
+		defer errors2.Capture(&rErr, func() error {
+			if rErr != nil {
+				// Close the new file to avoid leaks.
+				file.Close()
+			}
+			return rErr
+		})()
 		if err != nil {
-			// TODO: Close index if write fails.
+			p.logger.Error("manifest file write failed compacting index", zap.String("path", p.ManifestPath()), zap.Error(err))
```

Comment:
We should always either return an error (with appropriate context for the caller) or log the error, but not both. Otherwise we end up logging errors at every level of the call stack.

We can avoid being religious about that for now if necessary, but any time we log an error and then return the error is a code smell for me.